File: 2020/julia/src/Day3.jl
module Day3

include("Aoc.jl")
using .Aoc

map = Aoc.input_lines(3)

width = length(map[1])
height = size(map)[1]

isTree(pos) = map[pos[2]][((pos[1]-1) % width) + 1] == '#'

function treeCount(direction)
    pos = [1, 1]
    trees = 0

    while pos[2] <= height
        trees += isTree(pos) ? 1 : 0
        pos += direction
    end
    
    trees
end

part1() = treeCount([3, 1])
part2() = mapreduce(treeCount, *, eachrow([ 1 1; 3 1; 5 1; 7 1; 1 2 ]))

end

File: 2020/typescript/src/days/day3.ts
import {Day} from "../aoc/day";
import {product} from "./helpers";

export class Day3 extends Day {
    day = (): number => 3;

    part1 = (): string => this.treeCount({x: 3, y: 1}).toString();
    part2 = (): string => [
        {x: 1, y: 1},
        {x: 3, y: 1},
        {x: 5, y: 1},
        {x: 7, y: 1},
        {x: 1, y: 2}
    ]
        .map(this.treeCount)
        .reduce(product)
        .toString();

    private width = (): number => this.input[0].length;
    private height = (): number => this.input.length;

    private isTree = (v: Vector) => (this.input[v.y]?.[v.x % this.width()] ?? '') === '#';

    private add = (a: Vector, b: Vector): Vector => ({x: a.x + b.x, y: a.y + b.y});

    private treeCount = (direction: Vector): number => {
        let trees = 0;
        let position: Vector = {x: 0, y: 0};

        while (position.y < this.height()) {
            trees += Number(this.isTree(position));
            position = this.add(position, direction);
        }

        return trees;
    }
}

interface Vector {
    x: number;
    y: number;
}